torch/distributed/_cute_layout.py

```diff
@@ -0,0 +1,313 @@
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+# this list of conditions and the following disclaimer in the documentation
+# and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+#################################################################################################
+
+from functools import reduce
+from itertools import product
+
+
+class _Layout:
+    """
+    Utility class for representing a integer layouts by leveraging CuTe Layout Algebra.
+    See https://docs.nvidia.com/cutlass/media/docs/cpp/cute/02_layout_algebra.html for more details.
+
+    Each layout is represented as a list of (size, stride) pairs. We use it as a way for mechanical bookkeeping
+    of the integers such as ranks in a SPMD mesh, and the transformation on top of it.
+
+    Lots of methods of layout like coalesce, composition, complement, etc. are borrowed from pycute.
+    https://github.com/NVIDIA/cutlass/blob/6dd13d42784ee5bfa232d2441e6b9a021c5c6290/python/pycute/layout.py#L137,L257
+    """
+
+    def __init__(self, sizes_and_strides: list[tuple[int, int]]):
+        self.sizes_and_strides = sizes_and_strides
+
+    def __repr__(self) -> str:
+        return (
+            f"{self.__class__.__name__}(sizes_and_strides={self.sizes_and_strides!r})"
+        )
+
+    def __str__(self) -> str:
+        return f"Sizes/Strides: {self.sizes_and_strides}"
+
+    @property
+    def sizes(self) -> list[int]:
+        return [s for s, _ in self.sizes_and_strides]
+
+    @property
+    def strides(self) -> list[int]:
+        return [s for _, s in self.sizes_and_strides]
+
+    def numel(self) -> int:
+        return reduce(lambda x, y: x * y, self.sizes, 1)
+
+    def __eq__(self, other: object) -> bool:
+        return (
+            isinstance(other, _Layout)
+            and self.sizes_and_strides == other.sizes_and_strides
+        )
+
+    def __hash__(self) -> int:
+        return hash(tuple(self.sizes_and_strides))
```

Comment:
This seems to justify my earlier comment. We should make this `sizes_and_strides` a tuple?